File: 231123/xor 결과 최대 만들기/max-of-xor.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Main {
    private static final Scanner sc = new Scanner(System.in);

    private static int n, m, answer = Integer.MIN_VALUE;
    private static int[] numbers;

    public static void main(String[] args) {
        n = sc.nextInt();
        m = sc.nextInt();
        numbers = new int[n];
        for (int i = 0; i < n; i++) {
            numbers[i] = sc.nextInt();
        }

        recursive(new ArrayList<>(), 0);

        System.out.println(answer);
    }

    private static void recursive(List<Integer> nums, int idx) {
        if (nums.size() == m) {
            int result = nums.get(0);
            for (int i = 1; i < nums.size(); i++) {
                result ^= nums.get(i);
            }
            answer = Math.max(answer, result);
        }

        for (int i = idx; i < n; i++) {
            nums.add(numbers[i]);
            recursive(nums, idx + 1);
            nums.remove(nums.size() - 1);
        }
    }
}

File: 240506/xor 결과 최대 만들기/max-of-xor.kt
import java.util.*
import kotlin.math.max

val sc = Scanner(System.`in`)
val n = sc.nextInt()
val m = sc.nextInt()
val nums = mutableListOf<Int>()
var max = Int.MIN_VALUE

fun main() {
    for (i in 1..n) {
        nums.add(sc.nextInt())
    }

    recursive(0, mutableListOf())

    print(max)
}

fun recursive(
    idx: Int,
    selectedNums: MutableList<Int>
) {
    if (idx == n) {
        if (selectedNums.size == m) {
            var result = selectedNums[0]
            for (i in 1..selectedNums.size - 1) {
                result = result.xor(selectedNums[i])
            }
            max = max(max, result)
        }
        return
    }

    selectedNums.add(nums[idx])
    recursive(idx + 1, selectedNums)
    selectedNums.removeAt(selectedNums.size - 1)
    recursive(idx + 1, selectedNums)
}